src/persist-client/src/internal/apply.rs

```diff
@@ -44,12 +155,16 @@ pub struct Applier<K, V, T, D> {
     pub(crate) shard_metrics: Arc<ShardMetrics>,
     pub(crate) state_versions: Arc<StateVersions>,
 
-    // TODO: Wrap this in an Arc<tokio::sync::Mutex<_>> and have a single shared
-    // one for each shard in a process.
-    //
-    // NB: This is very intentionally not pub(crate) so that it's easy to reason
-    // very locally about the duration of Mutex holds.
-    state: TypedState<K, V, T, D>,
+    // Access to the shard's state, shared across all handles created by the same
+    // PersistClientCache. The state is wrapped in a std::sync::Mutex, disallowing
+    // access across await points. Access should be always be kept brief, and it
+    // is expected that other handles may advance the state at any time this Applier
+    // is not holding the lock.
+    state: Arc<Mutex<TypedState<K, V, T, D>>>,
+    // An Applier-local cache of several of the most frequently accessed (and cheapest
+    // to copy) values of the latest state, used to reduce lock contention. The cached
+    // state should always be updated after attempting to update the shared state.
+    cached_state: CachedState<T>,
```

Comment:
currently undecided if the complexity of the cached_state is worth it... it's kind of nice to have a consistent view of the cached state without locks, but I'm not sure if it provides a ton of value in practice